### DISCUSSION TITLE
Cannot query field "_service" on type "Query". I'm not using Federation.

### DISCUSSION THREAD

**Discussion by vmarchesin** [apollographql/apollo-server#6238] [Q&A] [answered]
So I have a GraphQL server built with apollo-server and recently I started to get this error. As far as I am aware I did not get this error before and I didn't make any changes to the schema that would create this error, I only added new functionality (new definitions) to the schema.

```bash
backend  | [1647789666605] ERROR (35 on 67ad11a8ae06): Cannot query field "_service" on type "Query".
backend  |     err: {
backend  |       "type": "ValidationError",
backend  |       "message": "Cannot query field \"_service\" on type \"Query\".",
backend  |       "locations": [
backend  |         {
backend  |           "line": 2,
backend  |           "column": 9
backend  |         }
backend  |       ],
backend  |       "extensions": {
backend  |         "code": "GRAPHQL_VALIDATION_FAILED",
backend  |         "exception": {
backend  |           "stacktrace": [
backend  |             "GraphQLError: Cannot query field \"_service\" on type \"Query\".",
backend  |             "    at Object.Field (/node_modules/graphql/validation/rules/FieldsOnCorrectTypeRule.js:48:31)",
backend  |             "    at Object.enter (/node_modules/graphql/language/visitor.js:323:29)",
backend  |             "    at Object.enter (/node_modules/graphql/utilities/TypeInfo.js:370:25)",
backend  |             "    at visit (/node_modules/graphql/language/visitor.js:243:26)",
backend  |             "    at validate (/node_modules/graphql/validation/validate.js:69:24)",
backend  |             "    at validate (/node_modules/apollo-server-core/src/requestPipeline.ts:470:27)",
backend  |             "    at processGraphQLRequest (/node_modules/apollo-server-core/src/requestPipeline.ts:261:30)",
backend  |             "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
backend  |             "    at async processHTTPRequest (/node_modules/apollo-server-core/src/runHttpQuery.ts:346:24)"
backend  |           ]
backend  |         }
backend  |       }
backend  |     }
```

A few things about this issue:

- This only happens when I open Apollo Studio. I do not get this error when interacting with the API, not even in Apollo Studio itself (the error is thrown once when opening it, any subsequent calls do not throw this error and work just fine).
- I am not using federation, as far as I know. I have only one schema that is using [@graphql-tools merge](https://www.graphql-tools.com/docs/schema-merging), in which it clearly states _This should not be confused with schema stitching, which builds a combined proxy schema atop numerous subservice APIs._ I do not have a gateway, nor am I accessing external GraphQL schemas or introspecting anything.

My Apollo Server config is as follows (I am using `apollo-server-express`):
```js
  const apolloServer = new ApolloServer({
    schema: makeExecutableSchema({
      typeDefs,
      resolvers,
      resolverValidationOptions: {
        requireResolversForResolveType: 'ignore',
      },
    }),
    introspection: true,
    playground: isDev,
    tracing: isDev,
    // Context is passed down to every resolver
    async context({ req }): Promise<AppContext> {
      const requestOwnerId = req.headers['authenticated-user'] as string;
      return { prisma, requestOwnerId };
    },
    formatError: errorFormatter, // error formatting function
    plugins: [
      ApolloServerPluginDrainHttpServer({ httpServer }),
    ],
  } as Config<ExpressContext>);
```

Every single instance of this error I found online mention federation, and I am confused as I am not using a federated schema whatsoever. What am I doing wrong? In Apollo Studio I can confirm I have a single schema, and not subgraphs. Thanks a lot.

**Comment by haykerman** [comment]
@vmarchesin Did you by any chance start using Apollo Studio recently? It seems the error is caused by Apollo Studio polling your schema.

**Comment by vmarchesin** [comment, marked as answer]
@haykerman I've been using Apollo Studio since I started with this project. I did not get any errors until recently, and I have no idea what could cause this besides what is in the federation docs. I just stripped my typedefs and resolvers to be as barebones as possible, and I just have the typedef and resolver for:

```js
// resolver
export default {
  Node: {
    __resolveType: (obj) =>  obj,
  },
};
```
```graphql
# typedef
scalar Cursor
scalar JSON
scalar JSONObject

interface Node {
  id: ID!
}

type Records {
  count: Int
}

type PageInfo {
  startCursor: String
  endCursor: String
}

type Query {
  node(id: ID!): Node
}

input PaginationInput {
  skip: Int
  take: Int
  cursor: Cursor
}

input FilterInput {
  orderBy: String
  sortOrder: JSON
}
```

Of course none of the queries work because I removed all the other stuff from the project, but just with this the server is able to start successfully and I still get the same error when loading Apollo Studio.

I'm using `@graphql-tools` to merge the typedefs and resolvers, but as I said I've been doing it since the beginning and I haven't changed, and only now been getting this error.

```js
// typeDefs/index.js
import { join } from 'path';
import { loadFilesSync } from '@graphql-tools/load-files';
import { mergeTypeDefs } from '@graphql-tools/merge';
import { DocumentNode } from 'graphql';

const typesArray = loadFilesSync(join(__dirname, '.'), { extensions: ['graphql'] });

export default mergeTypeDefs(typesArray) as DocumentNode;

// resolvers/index.js
import { mergeResolvers } from '@graphql-tools/merge';
import Common from './common';
import MoreResolvers from './otherResolverFiles';
// ... more resolver imports here
import AppContext from 'server/types/appContext';

export default mergeResolvers<any, AppContext>([
  Common,
  MoreResolvers,
  // ... more resolvers here
]);
```

These are imported as `typeDefs` and `resolvers` in the `new ApolloServer` that I mentioned above, here:
```js
  const apolloServer = new ApolloServer({
    schema: makeExecutableSchema({
      typeDefs,
      resolvers,
      resolverValidationOptions: {
        requireResolversForResolveType: 'ignore',
      },
    }),
  ...
```

So no schema stitching, no federation. It does seem like an Apollo Studio issue because I do not get the `Cannot query field "_service" on type "Query"` error when making a call to the GraphQL endpoint (Postman, Frontend App, even Apollo Studio itself after it loads) but I haven't found anything anywhere because 100% of the instances I found of people mentioning this error mention federation or subgraphs in one way or another. I might be doing something wrong, but I have no idea where to start looking because the only reasonable thing that could cause this problem (to my knowledge) is something I'm not using. I'm not incredibly GraphQL savvy so there's that. Thanks for taking the time to reply though.

**Reply by haykerman** [reply]
Seems like Apollo Studio assumes we use federation by default... But I am not sure. 
Those details you provided will be helpful to debug the issue for sure.

**Reply by glasser** [reply]
Hi, I'm looking into whether there were any recent changes in the Studio Explorer/Sandbox related to this. My guess is that when using Sandbox, Explorer uses a query like `{ _service { sdl } }` to decide whether to treat your service as a subgraph or not.

I assume this is an issue more in development than in production? After all, in production anyone who can access your graph can send invalid operations to you and so treating this as a error might not be appropriate.

**Reply by vmarchesin** [reply]
Just to clarify, this created zero problems for me other than this error popping up in the console as I open Apollo Studio in my browser. All queries work normally, as far as I've tested. Just sharing in hopes of finding a solution in case there's a problem with this that I haven't faced yet, errors are usually a bad thing to ignore.

**Reply by glasser** [reply]
OK, I think this should be fixed in Explorer now — it'll only try this query if your schema indicates that the field actually exists. (Note that to make sure you're on the latest version of the Explorer, you may need to click the bullhorn icon in the top right and click "Update available, click to relaunch" in the box that pops up.)

**Reply by vmarchesin** [reply]
I just tried it (Explorer says _You're up to date on version d857be_ after prompting an update) and I am not getting this error anymore. Definitely not something I did as I didn't make any changes since I opened this thread. So it look like it is fixed :)

If it pops up again I'll mention it here, thank you all for looking into it.

**Comment by vmarchesin** [comment]
I should mention that I am using the following dependencies (all updated to the latest available version):
```json
"dependencies": {
  "@graphql-tools/load-files": "^6.5.3",
  "@graphql-tools/merge": "^8.2.4",
  "@graphql-tools/schema": "^8.3.3",
  "@prisma/client": "^3.11.0",
  "apollo-server-core": "^3.6.4",
  "apollo-server-errors": "^3.3.1",
  "apollo-server-express": "^3.6.4",
  "body-parser": "^1.19.2",
  "dotenv": "^16.0.0",
  "express": "^4.17.3",
  "graphql": "^16.3.0",
  "graphql-type-json": "^0.3.2",
  "pino": "^7.9.1",
  "pino-pretty": "^7.5.4"
},
```

**Comment by eurbs** [comment]
By any chance have there been any recent changes to Apollo Studio that may have reintroduced the `Cannot query field "_service" on type "Query"` issue? I started running into this today, and the last time I successfully fetched a schema using Studio was earlier this week.

Same as above, this is only a problem when fetching the schema during development. I've looked into other clients for fetching the schema (e.g. Altair GraphQL Client) and the same issue is not present.

**Reply by trevor-scheer** [reply]
I can reproduce this (logging errors from a plugin)
```ts
  plugins: [
    {
      requestDidStart: async() => {
        return {
          didEncounterErrors({ errors }) {
            console.log(errors);
          }
        }
      }
    }
  ]
```
I've passed this feedback along to the team. Thanks for reporting!

**Reply by mayakoneval** [reply]
We have a solution up for review for this, thank you for reporting!

**Comment by mayakoneval** [comment]
Hello, a fix should have gone out for this a couple days ago. Let me know if you see this as a continued issue. Thanks for the report!

**Reply by eurbs** [reply]
Thanks for the fix @mayakoneval! It works perfectly as expected.
